dataplane/src/pipeline.rs

```diff
@@ -1,173 +1,168 @@
+        self.nf2.process(self.nf1.process(input))
     }
+}
 
-    fn update_config(&mut self, _: &Config) -> Result<(), Box<dyn Error>> {
-        Ok(())
-    }
+pub trait Pipeline<Buf: PacketBufferMut>: NetworkFunction<Buf> {
+    fn add_stage<NF: NetworkFunction<Buf>>(self, nf: NF) -> impl NetworkFunction<Buf>;
+}
 
-    fn process(
-        &mut self,
-        packets: Box<dyn Iterator<Item = MetaPacket>>,
-    ) -> Box<dyn Iterator<Item = MetaPacket>> {
-        packets
+impl<Buf: PacketBufferMut, Nf: NetworkFunction<Buf>> Pipeline<Buf> for Nf {
+    fn add_stage<NF: NetworkFunction<Buf>>(self, nf: NF) -> impl NetworkFunction<Buf>
+    where
+        Self: Sized,
+    {
+        Chain {
+            nf1: self,
+            nf2: nf,
+            _marker: PhantomData,
+        }
     }
 }
 
-// #[cfg(test)]
-// FIXME(mvachhar) remove this once we have a proper fake interface for packet/mbuf
-#[cfg(none)]
-#[allow(clippy::unwrap_used, clippy::panic, clippy::expect_used)]
+#[allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
+#[cfg(test)]
 mod test {
-    use super::{MetaPacket, Metadata, Passthrough, Pipeline};
-    use dpdk::mem::Mbuf;
-    use net::eth::ethertype::EthType;
+    use crate::packet::Packet;
+    use crate::pipeline::NetworkFunction;
+    use crate::pipeline::{BroadcastMacs, DecrementTtl, InspectHeaders, Pipeline};
+    use net::buffer::TestBuffer;
+    use net::eth::ethtype::EthType;
     use net::eth::mac::{DestinationMac, Mac, SourceMac};
     use net::eth::Eth;
-    use net::packet::Packet;
+    use net::headers::{Headers, Net, TryEth};
+    use net::ipv4::addr::UnicastIpv4Addr;
+    use net::ipv4::Ipv4;
+    use net::parse::DeParse;
+    use std::net::Ipv4Addr;
 
     #[test]
-    fn test_passthrough_process() {
-        let mut pipeline = Pipeline::new();
-        let passthrough = Passthrough::new();
-        pipeline.add_stage(Box::new(passthrough)).unwrap();
-
-        let src_mac = SourceMac::new(Mac::from([0x00, 0x00, 0x00, 0x00, 0x00, 0x01])).unwrap();
-        let dst_mac = DestinationMac::new(Mac::from([0x00, 0x00, 0x00, 0x00, 0x00, 0x02])).unwrap();
-        let test_packet = MetaPacket {
-            packet: Packet::new(Eth::new(src_mac, dst_mac, EthType::IPV4)),
-            metadata: Metadata { vni: None },
-            outer_packet: None,
-            // Temporary mbuf fake, need to have a proper mockable interface
-            #[allow(unsafe_code)]
-            mbuf: unsafe {
-                Mbuf::new_from_raw_unchecked(std::ptr::null_mut::<dpdk_sys::rte_mbuf>().offset(1))
-            },
-        };
-
-        let input_packets = vec![test_packet];
-        let output: Vec<_> = pipeline
-            .process_packets(Box::new(input_packets.into_iter()))
-            .collect();
-
-        assert_eq!(1, output.len());
-        assert_eq!(src_mac, output[0].packet.eth.source());
-        assert_eq!(dst_mac, output[0].packet.eth.destination());
+    fn biscuit() {
```

Comment:
Can we get meaningful function names, or is this scratch code that should be dropped?

Reply:
For sample code we should write some docs and have the docs get tested using the doc-test stuff, which should run with cargo test.

Reply:
I agree in principal but the point is moot here since this whole commit is dropped now